File: desktop/tests/Qlipq.Ffmpeg.Tests/ProgressTests.cs
using Qlipq.Ffmpeg;
using Xunit;

namespace Qlipq.Ffmpeg.Tests;

public class ProgressTests
{
    [Fact]
    public void ParseTimecodeConvertsToSeconds()
    {
        Assert.Equal(1.5, Progress.ParseTimecode("00:00:01.500000"));
        Assert.Equal(3723, Progress.ParseTimecode("01:02:03"));
        Assert.Null(Progress.ParseTimecode("nope"));
    }

    [Fact]
    public void ParseProgressReadsOutTimeUsAndContinueState()
    {
        var chunk = string.Join("\n", "frame=120", "out_time_us=2500000", "out_time=00:00:02.500000", "progress=continue");
        var result = Progress.ParseProgress(chunk);
        Assert.Equal(2.5, result.OutTimeSec);
        Assert.False(result.Done);
    }

    [Fact]
    public void ParseProgressDetectsEndMarker()
    {
        var result = Progress.ParseProgress("out_time_us=10000000\nprogress=end\n");
        Assert.Equal(10, result.OutTimeSec);
        Assert.True(result.Done);
    }

    [Fact]
    public void ParseProgressFallsBackToOutTimeTimecode()
    {
        var result = Progress.ParseProgress("out_time=00:00:04.000000\nprogress=continue");
        Assert.Equal(4, result.OutTimeSec);
    }

    [Fact]
    public void ProgressFractionClampsToUnitInterval()
    {
        Assert.Equal(0.5, Progress.ProgressFraction(5, 10));
        Assert.Equal(1, Progress.ProgressFraction(20, 10));
        Assert.Equal(0, Progress.ProgressFraction(null, 10));
        Assert.Equal(0, Progress.ProgressFraction(5, 0));
    }
}

File: iced/crates/qlipq-ffmpeg/tests/progress_tests.rs
use qlipq_ffmpeg::progress::*;

fn close(a: f64, b: f64) -> bool {
    (a - b).abs() < 1e-9
}

#[test]
fn parse_timecode_to_seconds() {
    assert!(close(parse_timecode("00:00:01.500000").unwrap(), 1.5));
    assert!(close(parse_timecode("01:02:03").unwrap(), 3723.0));
    assert_eq!(parse_timecode("nope"), None);
}

#[test]
fn parse_progress_reads_out_time_us_and_continue() {
    let chunk = "frame=120\nout_time_us=2500000\nout_time=00:00:02.500000\nprogress=continue";
    let r = parse_progress(chunk);
    assert!(close(r.out_time_sec.unwrap(), 2.5));
    assert!(!r.done);
}

#[test]
fn parse_progress_detects_end() {
    let r = parse_progress("out_time_us=10000000\nprogress=end\n");
    assert!(close(r.out_time_sec.unwrap(), 10.0));
    assert!(r.done);
}

#[test]
fn parse_progress_falls_back_to_timecode() {
    let r = parse_progress("out_time=00:00:04.000000\nprogress=continue");
    assert!(close(r.out_time_sec.unwrap(), 4.0));
}

#[test]
fn progress_fraction_clamps() {
    assert!(close(progress_fraction(Some(5.0), 10.0), 0.5));
    assert!(close(progress_fraction(Some(20.0), 10.0), 1.0));
    assert!(close(progress_fraction(None, 10.0), 0.0));
    assert!(close(progress_fraction(Some(5.0), 0.0), 0.0));
}
